src/rmp/src/Restructure.cpp

```diff
@@ -0,0 +1,555 @@
+  }
+  files_to_remove.emplace_back(abc_script_file);
+  if (best_inst_count < std::numeric_limits<int>::max()) {
+    // read back netlist
+    debugPrint(logger_, RMP, "remap", 1, "Reading blif file {}", best_blif);
+    blif_.readBlif(best_blif.c_str(), block_);
+    debugPrint(logger_,
+               utl::RMP,
+               "remap",
+               1,
+               "Number constants after restructure {}",
+               countConsts(block_));
+  }
+
+  for (auto file_to_remove : files_to_remove) {
+    if (!logger_->debugCheck(RMP, "remap", 16))
+      std::remove(file_to_remove.c_str());
+  }
+}
+
+void Restructure::postABC(float worst_slack)
+{
+  rsz::Resizer* resizer = openroad_->getResizer();
+  bool area_mode = opt_mode_ <= Mode::AREA_3;
+  if (!area_mode) {
+    // Recompute timing
+    resizer->estimateWireParasitics();
+    open_sta_->findRequireds();
+    float new_slack;
+    sta::Vertex* worst_vertex;
+    open_sta_->worstSlack(sta::MinMax::max(), new_slack, worst_vertex);
+    if (!area_mode && new_slack < worst_slack) {
+      // failed, revert
+
+    } else {
+      // slack improved, accept
+    }
+  }
+
+  // Leave the parasitices up to date.
+  resizer->estimateWireParasitics();
+}
+void Restructure::getEndPoints(sta::PinSet& ends,
+                               bool area_mode,
+                               unsigned max_depth)
+{
+  open_sta_->ensureGraph();
+  open_sta_->searchPreamble();
+  auto sta_state = open_sta_->search();
+  int path_count = 100000;
+  float min_slack = area_mode ? -sta::INF : -sta::INF;
+  float max_slack = area_mode ? sta::INF : 0;
+
+  sta::PathEndSeq* path_ends
+      = sta_state->findPathEnds(  // from, thrus, to, unconstrained
+          nullptr,
+          nullptr,
+          nullptr,
+          false,
+          // corner, min_max,
+          open_sta_->findCorner("default"),
+          sta::MinMaxAll::max(),
+          // group_count, endpoint_count, unique_pins
+          path_count,
+          2,
+          true,
+          min_slack,
+          max_slack,  // slack_min, slack_max,
+          true,       // sort_by_slack
+          nullptr,    // group_names
+          // setup, hold, recovery, removal,
+          true,
+          true,
+          false,
+          false,
+          // clk_gating_setup, clk_gating_hold
+          false,
+          false);
+  bool skip_top_paths = false;
+  std::size_t path_found = path_ends->size();
+  if (path_found > 1000) {
+    skip_top_paths = true;
+  }
+  logger_->report("Number of paths for restructure {}", path_found);
+  int end_count = 0;
+  for (auto& path_end : *path_ends) {
+    if (opt_mode_ >= Mode::DELAY_1) {
+      sta::PathExpanded expanded(path_end->path(), open_sta_);
+      logger_->report("Got path of depth {}", expanded.size() / 2);
+      if (expanded.size() / 2 > max_depth) {
+        ends.insert(path_end->vertex(sta_state)->pin());
+        end_count++;
+      }
+      if (end_count > 5)  // limit blob size for timing
```

Comment:
why 5?

Reply:
Upto 5 size should be ok. We may have 4 as well. This would become clear once we experiment with timing restructuring on few designs.